From e1d37ab2519ac7652456f6512dbab1ed7180f3b0 Mon Sep 17 00:00:00 2001
From: Aleix Pol <aleixpol@kde.org>
Date: Wed, 24 Jan 2024 01:54:43 +0100
Subject: [PATCH] kio-open: Special-case thunderlinks

They often are shipped with 2 // making the payload of the links the
hostname which, per-spec, is case insensitive and thus QUrl leverages
this property.
This breaks thunderlinks because they encode information in the link
that is case sensitive.
If we put it at the path section, then they seem to work fine.

CCBUG: 429408
---
 kioclient/kioclient.cpp | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/kioclient/kioclient.cpp b/kioclient/kioclient.cpp
index 98f7cc0c..69f687a2 100644
--- a/kioclient/kioclient.cpp
+++ b/kioclient/kioclient.cpp
@@ -232,8 +232,16 @@ static void checkArgumentCount(int count, int min, int max)
 }
 
 #ifndef KIOCORE_ONLY
-bool ClientApp::kde_open(const QString &url, const QString &mimeType, bool allowExec)
+bool ClientApp::kde_open(const QString &_url, const QString &mimeType, bool allowExec)
 {
+    QString url = _url;
+    if ((url.startsWith(QLatin1String("thunderlink://"))) || (url.startsWith(QLatin1String("slack://"))) || (url.startsWith(QLatin1String("zoommtg://")))) {
+        // :// -> :/// so QUrl doesn't break the payload's
+        // case sensitivity which would otherwise be at the hostname section
+        const QLatin1Char slash('/');
+        url.insert(url.indexOf(slash), slash);
+    }
+
     UrlInfo info(url);
 
     if (!info.atStart()) {
-- 
GitLab

